File: solutions/0906-walking-robot-simulation/solution.js
/**
 * @param {number[]} commands
 * @param {number[][]} obstacles
 * @return {number}
 */
var robotSim = function(commands, obstacles) {
    let position = [0,0];    
    let max = 0;
    const rocks = new Map();
    
    const vectors = [
        [0,1], //n
        [1,0], //e
        [0,-1],//s
        [-1,0] //w
    ]
    
    let currentDir = 0;
    
    for(const rock of obstacles){
        if(!rocks[rock[0]]){
            rocks[rock[0]] = new Set();
        }
        rocks[rock[0]].add(rock[1]);
    };
    
    for(const command of commands){
        if(command === -1){
            currentDir = currentDir + 1
            currentDir %= 4;
        }else if (command === -2){
            currentDir = currentDir - 1;
            if(currentDir === -1){
                currentDir = 3;
            }
        }else{
            for(let i = 0; i < command; i++){
                const nextX = position[0] + vectors[currentDir][0];
                const nextY = position[1] + vectors[currentDir][1];
                if(rocks[nextX] && rocks[nextX].has(nextY)){
                    break;
                }else{
                    position = [nextX, nextY];
                    max = Math.max(position[0]**2 + position[1] **2, max);
                }
            }   
        }
    }

    return max;
};

File: solutions/0906-walking-robot-simulation/solution.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        s = set()
        for o in obstacles:
            s.add(tuple(o))
        pos = [0,0]

        dirs = [[0,1],[1,0],[0,-1],[-1,0]]
        current_dir = 0

        max_dist = 0

        for c in commands:
            if c == -2:
                current_dir -= 1
                current_dir %= 4
            elif c == -1:
                current_dir += 1
                current_dir %= 4
            else:
                for i in range(c):
                    pos[0] += dirs[current_dir][0]
                    pos[1] += dirs[current_dir][1]
                    if tuple(pos) in s:
                        pos[0] -= dirs[current_dir][0]
                        pos[1] -= dirs[current_dir][1]
                    max_dist = max(max_dist, pos[0] ** 2 + pos[1] ** 2)

        return max_dist
